pkg/kubelet/volume/cache/actual_state_of_world.go

```diff
@@ -265,7 +269,42 @@ type mountedPod struct {
 
 func (asw *actualStateOfWorld) MarkVolumeAsAttached(
 	volumeSpec *volume.Spec, nodeName string, devicePath string) error {
-	_, err := asw.AddVolume(volumeSpec, devicePath)
+	volumeName, err := asw.AddVolume(volumeSpec, devicePath)
+	if err != nil {
+		return err
+	}
+	if asw.kubeClient == nil {
+		return nil
+	}
+	// Update the node status in the apiserver because this will prevent
+	// the attach-detach controller from mutating the same volume.
+	// For the same reason, if this fails don't retry, because the
+	// attache-detach controller might've taken the lock. It could fail
+	// for different reasons, eg: kubelet status update; returning an error
+	// should cause the caller to retry.
+	if err := asw.addVolumeToNode(volumeName, nodeName); err != nil {
+		// Rollback the AddVolume above, if we can't record it on the node.
+		asw.MarkVolumeAsDetached(volumeName, nodeName)
+		return err
+	}
+	return nil
+}
+
+func (asw *actualStateOfWorld) addVolumeToNode(volumeName api.UniqueVolumeName, nodeName string) error {
+	node, err := asw.kubeClient.Core().Nodes().Get(nodeName)
```

Comment:
Maybe we should just not bother with qps, document that it might be high with volumes and move on with 1.3?
Also, i suppose this will update even for non-gce volumes.